reps/2022-09-19-ray-on-spark.md

```diff
@@ -0,0 +1,258 @@
+## Summary
```

Comment:
For NodeProvider approach, I think we need to do some prototype in order to better understand it. Based on my current understanding, the pros and cons are:

Pros:
    1.  The same and consistent way to config nodes via yaml. For example we can specify different custom resources for each node type. We can also config docker for Ray nodes. (Those are not possible with the current `init_cluster()` api so it's more future looking).
    2.  Supports autocaling. (This is also something not supported by current proposal).
    
  Cons:
       1. The current NodeProvider abstraction doesn't match our model very well. For example, NodeProvider separates getting physical node and starting ray node while this design does both via a single spark barrier job. Also NodeProvider api creates one node at a time while ray-on-spark needs to create the whole cluster at once. I think it's still doable but it will be more complicated than the `init_cluster()` approach due to the need to conform to the NodeProvider api.
       
While NodeProvider approach might be more future-proof, the benefit probably doesn't justify the extra complexity for now and it has more unknowns without a prototype. We can probably just start with `init_cluster()` approach for now. In the future, we can always switch to the NodeProvider approach if needed.

Thoughts @ericl @scv119 @WeichenXu123 ?

Reply:
We agree that a NodeProvider approach is appealing and definitely something to strive towards if users struggle with the barrier execution mode. We're a bit reluctant to move down that path initially though, mostly for the cons that you've listed so well.
In the interests of time and effort, implementing even a prototype would likely be leveraging internal Databricks functionality to get auto-scaling to control the initialization of worker nodes, perform health checks, and restart dropped nodes. We'd like to avoid such an implementation since many of those features are not part of OSS Spark and we'd like to avoid restricting users to the Databricks platform to use this integration. 
However, we'd still like to think about how to extend this functionality and properly scope this work after an initial best-effort implementation that utilizes barrier execution mode is done. If the need arises to build such advanced functionality, we can collaborate on a design and implementation that serves this use case in the future. 

Additional thoughts and comments @ericl @scv119 @dbczumar @WeichenXu123 ?

Reply:
@BenWilson2 I completely agree. Echoing the assessment from @jjyao and @ericl, it seems very reasonable to start with the existing proposal which is clearly OSS-compatible and investigate a node provider solution in the coming months. I concur that there are some difficult questions that would need to be answered in order to integrate Ray's node provider abstraction with Spark. @ericl @scv119 @zhe-thoughts @jjyao do we have alignment here?

Reply:
We can support NodeProvider in next step. But I think databricks spark cluster might be incompatible with NodeProvider.

Reply:
sgtm.